File: backend/app/libraries/engine.py
"""Investigation engine (mirrors investigation/engine.ts)."""
import asyncio
import logging
import secrets

from .. import db, config
from ..libraries import gateway as gateway_mod
from ..libraries.analyze import heuristic_analysis, build_evidence_graph
from ..libraries import analyzers

logger = logging.getLogger(__name__)

USDC_DECIMALS = config.ALGORAND_USDC_DECIMALS

RESOLVED_STATES = [
    "evidence_received", "settled", "evidence_unavailable", "provider_unavailable", "payment_failed",
    "settlement_failed", "wallet_rejected", "unsupported_network",
    "insufficient_balance", "evidence_request_failed",
]


def _nanoid(prefix: str) -> str:
    return f"{prefix}_{secrets.token_urlsafe(6)[:8]}"


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()


def detect_input_type(inputs: list[dict]) -> str:
    types = {i.get("type") for i in inputs}
    if len(types) > 1:
        return "mixed"
    return inputs[0]["type"] if inputs and inputs[0].get("type") else "text"


def _set_stage(inv, stage_id, status):
    for s in inv.get("stages") or []:
        if s["id"] == stage_id:
            s["status"] = status


def _init_stages() -> list[dict]:
    return [{"id": s["id"], "label": s["label"], "status": "pending"} for s in config.INVESTIGATION_STAGES]


def _emit(inv, kind, label, detail=None):
    inv.setdefault("activity", []).append({
        "id": _nanoid("evt"), "investigationId": inv["id"], "userId": inv.get("userId"),
        "kind": kind, "label": label, "detail": detail, "createdAt": _now_iso(),
    })


def start_capability_investigation(input_: dict) -> dict:
    cap = config.get_paid_capability(input_.get("capability") or "")
    inv = gateway_mod.create_investigation_record({
        "id": input_["id"], "userId": input_.get("userId"),
        "title": input_.get("title") or (cap["title"] if cap else "Investigation in Progress"),
        "question": input_["question"], "inputs": input_.get("inputs") or [],
        "inputType": detect_input_type(input_.get("inputs") or []),
        "status": "created",
        "capability": input_.get("capability"),
        "capabilityPriceUsdc": cap.get("priceUsdc") if cap else None,
        "idempotencyKey": input_.get("idempotencyKey"),
    })
    inv["stages"] = _init_stages()
    _emit(inv, "investigation_created", f"{(cap['title'] if cap else 'Capability')} investigation started")
    db.save_investigation(inv)
    return db.get_investigation(inv["id"])


def plan_capability(inv: dict, requirements: list[dict]) -> dict:
    inv2 = db.get_investigation(inv["id"])
    inv2["status"] = "planning"
    _set_stage(inv2, "planning", "active")
    db.save_investigation(inv2)
    inv2["evidenceRequirements"] = requirements or []
    inv2["investigationPlan"] = [
        {"id": r["id"], "capability": r["capability"], "reason": r["reason"],
         "estimatedCost": 0, "expectedValue": 0, "status": "pending"}
        for r in (requirements or [])
    ]
    _set_stage(inv2, "planning", "completed")
    inv2["status"] = "discovering"
    inv2["updatedAt"] = _now_iso()
    db.save_investigation(inv2)
    return db.get_investigation(inv2["id"])


async def await_finalize_investigation(id, analyze=None, allow_input_analysis=False) -> dict:
    """Async wrapper around finalize (analyzers are async)."""
    inv = db.get_investigation(id)
    if not inv:
        raise ValueError("Investigation not found")

    acqs = inv.get("acquisitions") or []
    unresolved = [a for a in acqs if a.get("paymentState") not in RESOLVED_STATES]
    if unresolved:
        return inv

    evidence = [e for e in (inv.get("evidence") or []) if e.get("status") == "collected"]
    analyzer = analyze or analyzers.get_analyzer(inv.get("capability")) or heuristic_analysis

    if not evidence and not allow_input_analysis:
        # Zero acquired external evidence → NO verdict, regardless of analyzer.
        # Capability analyzers exist for every capability; they must never run
        # on an empty evidence set (that produced fabricated assessments).
        inv["status"] = "evidence_unavailable"
        inv["conclusion"] = "insufficient_evidence"
        inv["conclusionText"] = (
            "Insufficient evidence was acquired to render an assessment. The investigation could not be "
            "completed because no paid external evidence was obtained."
        )
        inv["confidence"] = 0
        inv["limitations"] = [
            "No external evidence was successfully acquired",
            inv.get("blockReason") or "Evidence acquisition did not complete",
        ]
        inv["updatedAt"] = _now_iso()
        db.save_investigation(inv)
        return inv

    inv["status"] = "analyzing"
    _set_stage(inv, "analyzing", "active")
    db.save_investigation(inv)

    if asyncio.iscoroutinefunction(analyzer):
        result = await analyzer(inv, evidence)
    else:
        result = analyzer(inv, evidence)

    inv["status"] = "cross_checking"
    _set_stage(inv, "analyzing", "completed")
    _set_stage(inv, "cross_checking", "active")
    inv["supportingEvidenceIds"] = [
        e["id"] for e in evidence if e.get("signal") == "supporting" or e.get("supportsClaim")
    ]
    inv["contradictoryEvidenceIds"] = [
        e["id"] for e in evidence if e.get("signal") == "contradictory" or e.get("contradictsClaim")
    ]
    _emit(inv, "cross_check_completed", "Cross-check of acquired evidence completed")
    db.save_investigation(inv)

    inv["conclusion"] = result["conclusion"]
    inv["conclusionText"] = result["conclusionText"]
    inv["confidence"] = result["confidence"]
    inv["risk"] = result["risk"]
    inv["findings"] = result["findings"]
    inv["limitations"] = result["limitations"]
    inv["contradictions"] = result["contradictions"]
    inv["sourcesUsed"] = (
        result.get("sourcesUsed") if result.get("sourcesUsed") else [e.get("source") for e in evidence]
    )
    if not inv.get("evidenceGraph") or not (inv["evidenceGraph"].get("nodes") or inv["evidenceGraph"].get("edges")):
        inv["evidenceGraph"] = build_evidence_graph(inv, evidence)
    inv_payments = db.get_payments_for_investigation(inv["id"])
    capability_fee = sum(
        float(p.get("amount") or 0)
        for p in inv_payments if p.get("status") == "settled" and p.get("capability") == inv.get("capability")
    )
    inv["economicSummary"] = {
        "totalSpend": round(capability_fee, 6),
        "capabilityFeeUsdc": round(capability_fee, 6),
        "checksPurchased": len([a for a in acqs if a.get("evidence")]),
        "providerCategories": list(dict.fromkeys(a.get("capability") for a in acqs if a.get("evidence"))),
        "settlementStatus": "Settled",
        "algorandRef": next((a.get("txId") for a in acqs if a.get("txId")), None),
        "paymentReferences": [p.get("settlementRef") for p in inv_payments
                              if p.get("status") == "settled" and p.get("settlementRef")],
    }

    # Evidence reasoning: classify roles, duplicates, stale, independent,
    # missing types, and surface relationships in the final result.
    plan_types = {r.get("capability") or r.get("type")
                  for r in (inv.get("evidenceRequirements") or [])}
    acquired_types = {e.get("type") or e.get("capability") for e in evidence}
    missing_types = sorted(plan_types - acquired_types)
    classification = _classify_evidence(
        evidence, inv.get("duplicates"), missing_types)
    inv["evidenceClassification"] = classification

    # Aggregate relationships from provider metadata (never fabricated).
    rel_seen = set()
    rel_agg = []
    for e in evidence:
        for r in (e.get("relationships") or []):
            key = tuple(sorted(r.items())) if isinstance(r, dict) else str(r)
            if key not in rel_seen:
                rel_seen.add(key)
                rel_agg.append(r)

    inv["finalResult"] = {
        "decision": result["conclusion"],
        "conclusionText": result["conclusionText"],
        "confidence": result["confidence"],
        "supportingEvidenceIds": [e["id"] for e in evidence if e.get("signal") == "supporting"],
        "contradictoryEvidenceIds": [e["id"] for e in evidence if e.get("signal") == "contradictory"],
        "uncertainEvidenceIds": [e["id"] for e in evidence if e.get("signal") == "uncertain"],
        "evidenceCount": len(evidence),
        "relationships": rel_agg,
        "missingEvidenceTypes": missing_types,
        "paymentReferences": inv["economicSummary"]["paymentReferences"],
        "limitations": result.get("limitations") or [],
        "independenceNote": classification["independenceNote"],
        "assurance": "Assessment is evidence-based, not a guarantee of truth.",
    }

    inv["status"] = "completed"
    _set_stage(inv, "cross_checking", "completed")
    _set_stage(inv, "assessment", "completed")
    inv["currentStage"] = None
    inv["updatedAt"] = _now_iso()
    _emit(inv, "assessment_generated", "Assessment generated for capability investigation")
    db.save_investigation(inv)
    return inv


async def analyze_investigation(investigation_id: str) -> dict:
    """Analyze the user's submitted input with Inquvia's own analyzers."""
    inv = db.get_investigation(investigation_id)
    if not inv:
        raise ValueError("Investigation not found")

    _set_stage(inv, "discovering", "completed")
    _set_stage(inv, "evidence_requested", "completed")
    _set_stage(inv, "evidence_received", "completed")
    _emit(inv, "internal_analysis_started", "Analyzing submitted data with Inquvia tools")
    inv["status"] = "analyzing"

    # Create internal evidence entries from submitted inputs so the report can
    # display meaningful evidence count, source nodes in the graph, and the AI
    # analyzer has items to reason over instead of an empty evidence list.
    from ..libraries import storage as _storage
    existing_ev_sources = {e.get("metadata", {}).get("origin") for e in (inv.get("evidence") or [])}
    if "user_submission" not in existing_ev_sources:
        for inp in inv.get("inputs") or []:
            content_preview = inp.get("content") or inp.get("fileName") or "submitted input"
            text_content = None
            if inp.get("filePath"):
                text_content = _storage.read_stored_text(inp["filePath"], 500)
            finding_text = (text_content[:200].strip() if text_content else f"User-submitted {content_preview}")
            ev_item = {
                "id": _nanoid("ev"),
                "type": inp.get("type", "document"),
                "source": f"Submitted {inp.get('type', 'input')}: {content_preview}",
                "finding": finding_text,
                "signal": "uncertain",
                "status": "collected",
                "confidence": None,
                "timestamp": _now_iso(),
                "metadata": {
                    "origin": "user_submission",
                    "fileName": inp.get("fileName") or inp.get("content"),
                    "inputType": inp.get("type"),
                },
            }
            inv.setdefault("evidence", []).append(ev_item)

    db.save_investigation(inv)
    return await await_finalize_investigation(investigation_id, allow_input_analysis=True)



def _classify_evidence(evidence: list[dict], duplicates, missing_types: list[str]) -> dict:
    """Classify evidence roles without fabricating conclusions.

    signal and confidence come from the provider's EvidenceResponse; duplicate
    pairs come from the provider's duplicates analysis when present; stale is
    a local age check; independence is only asserted when a provider analysis
    says so (otherwise reported as unknown).
    """
    import json as _json
    from datetime import datetime, timezone

    pairs = {}
    try:
        raw = duplicates or {}
        pair_list = raw.get("duplicates") if isinstance(raw, dict) else raw
        for p in pair_list or []:
            if isinstance(p, (list, tuple)) and len(p) >= 2:
                pairs.setdefault(p[0], []).append(p[1])
                pairs.setdefault(p[1], []).append(p[0])
    except Exception:
        pairs = {}

    now = datetime.now(timezone.utc)
    items = []
    for e in evidence:
        ts = None
        try:
            ts = datetime.fromisoformat((e.get("timestamp") or "").replace("Z", "+00:00"))
        except Exception:
            ts = None
        stale = bool(ts and (now - ts).days > 30)
        conf = e.get("confidence") or 0
        items.append({
            "id": e["id"],
            "type": e.get("type"),
            "signal": e.get("signal") or "uncertain",
            "nonProbative": bool(conf < 20 or not e.get("finding")),
            "stale": stale,
            "duplicates": pairs.get(e["id"], []),
            "independent": e.get("independent") if e.get("independent") is not None else None,
            "source": e.get("source"),
            "relationships": e.get("relationships") or [],
        })

    any_duplicate = bool(pairs)
    return {
        "items": items,
        "missing": sorted(set(missing_types)),
        "duplicatePairsFound": any_duplicate,
        "independenceNote": (
            "Independence is counted only when explicitly declared by the provider; "
            "missing declarations remain unknown."
        ),
    }



async def run_investigation(request: dict) -> dict:
    id_ = request.get("id") or f"case_{secrets.token_urlsafe(6)[:10]}"
    inv = db.get_investigation(id_)
    if not inv:
        inv = gateway_mod.create_investigation_record({
            "id": id_, "userId": request.get("userId"), "title": "Investigation in Progress",
            "question": request.get("question", ""), "inputs": request.get("inputs") or [],
            "inputType": detect_input_type(request.get("inputs") or []), "status": "created",
        })
        inv["stages"] = _init_stages()
        db.save_investigation(inv)
        _emit(inv, "investigation_created", "Investigation created")

    if inv["status"] in ("created", "planning"):
        from ..libraries.planner import plan_evidence_requirements
        requirements = plan_evidence_requirements(
            inv.get("question", ""), [i.get("type") for i in (inv.get("inputs") or [])]
        )
        inv = plan_capability(inv, requirements)

    if inv["status"] == "discovering":
        inv = await analyze_investigation(id_)

    return db.get_investigation(id_)


def get_investigation_progress(id) -> dict | None:
    return db.get_investigation(id)

File: src/lib/investigation/engine.ts
import { nanoid } from "nanoid";
import {
  INVESTIGATION_STAGES,
  INVESTIGATION_BLOCKED_STATES,
  getPaidCapability,
} from "../config";
import {
  getInvestigation,
  saveInvestigation,
  getUserById,
} from "../db";
import type {
  ActivityEvent,
  AssessmentLabel,
  EvidenceItem,
  Investigation,
  InvestigationInput,
  InputType,
  PaymentState,
  RiskLevel,
  StageProgress,
} from "../types";
import { planEvidenceRequirements } from "./planner";
import {
  planAcquisitions,
  createInvestigationRecord,
} from "../gateway";
import { acquireDownstream } from "../gateway";
import { createServerSigner } from "../wallet/serverSigner";
import { AnalysisResult, heuristicAnalysis, buildEvidenceGraph } from "./analyze";
import { getAnalyzer, type CapabilityAnalyzer } from "./analyzers";

export interface InvestigationRequest {
  question: string;
  inputs: InvestigationInput[];
  id?: string;
  userId?: string;
}

/** Acquisition states that are finished (resolution can proceed past them). */
const RESOLVED_STATES: PaymentState[] = [
  "evidence_received",
  "settled",
  "provider_unavailable",
  "payment_failed",
  "settlement_failed",
  "wallet_rejected",
  "unsupported_network",
  "insufficient_balance",
  "evidence_request_failed",
];

function detectInputType(inputs: InvestigationInput[]): InputType {
  const types = new Set(inputs.map((i) => i.type));
  if (types.size > 1) return "mixed";
  return inputs[0]?.type ?? "text";
}

function setStage(inv: Investigation, stageId: string, status: StageProgress["status"]) {
  inv.stages = inv.stages.map((s) =>
    s.id === stageId ? { ...s, status } : s
  );
}

function initStages(): StageProgress[] {
  return INVESTIGATION_STAGES.map((s) => ({
    id: s.id,
    label: s.label,
    status: "pending" as const,
  }));
}

function emit(
  inv: Investigation,
  kind: ActivityEvent["kind"],
  label: string,
  detail?: string
) {
  inv.activity = inv.activity ?? [];
  inv.activity.push({
    id: `evt_${nanoid(8)}`,
    investigationId: inv.id,
    userId: inv.userId,
    kind,
    label,
    detail,
    createdAt: new Date().toISOString(),
  });
}

export interface CapabilityStartInput {
  id: string;
  userId?: string;
  question: string;
  inputs: InvestigationInput[];
  capability: string;
  title?: string;
  idempotencyKey?: string;
}

/**
 * Create the record for a new atomic capability investigation, recording which
 * x402 endpoint/capability was used and its price.
 */
export async function startCapabilityInvestigation(
  input: CapabilityStartInput
): Promise<Investigation> {
  const cap = getPaidCapability(input.capability);
  const inv = createInvestigationRecord({
    id: input.id,
    userId: input.userId,
    title: input.title ?? cap?.title ?? "Investigation in Progress",
    question: input.question,
    inputs: input.inputs,
    inputType: detectInputType(input.inputs),
    status: "created",
    capability: input.capability,
    capabilityPriceUsdc: cap?.priceUsdc,
    idempotencyKey: input.idempotencyKey,
  });
  inv.stages = initStages();
  emit(inv, "investigation_created", `${cap?.title ?? "Capability"} investigation started`);
  await saveInvestigation(inv);
  return (await getInvestigation(inv.id)) as Investigation;
}

/** Record the capability-specific evidence requirements and advance to discovery. */
export async function planCapability(inv: Investigation, requirements: Investigation["evidenceRequirements"]): Promise<Investigation> {
  const inv2 = (await getInvestigation(inv.id)) as Investigation;
  inv2.status = "planning";
  setStage(inv2, "planning", "active");
  await saveInvestigation(inv2);

  inv2.evidenceRequirements = requirements ?? [];
  inv2.investigationPlan = (requirements ?? []).map((r) => ({
    id: r.id,
    capability: r.capability,
    reason: r.reason,
    estimatedCost: 0,
    expectedValue: 0,
    status: "pending" as const,
  }));
  setStage(inv2, "planning", "completed");
  inv2.status = "discovering";
  inv2.updatedAt = new Date().toISOString();
  await saveInvestigation(inv2);
  return (await getInvestigation(inv2.id)) as Investigation;
}

/**
 * Advance an investigation through discovery and acquisition. When the server
 * wallet is configured, downstream evidence services are paid from Inquvia's
 * wallet and the investigation is finalized. Otherwise it enters the
 * awaiting_payment state for the client-side flow. Capability investigations
 * with nothing left to acquire are finalized immediately (the capability
 * analyzer runs on the submitted input itself).
 */
export async function discoverAndAcquire(
  investigationId: string,
  userId: string
): Promise<Investigation> {
  let inv2 = (await getInvestigation(investigationId)) as Investigation;
  if (!inv2) throw new Error("Investigation not found");

  setStage(inv2, "discovering", "active");
  await saveInvestigation(inv2);

  const plan = await planAcquisitions(investigationId, userId);
  inv2 = (await getInvestigation(investigationId)) as Investigation;
  inv2.acquisitions = plan.acquisitions;
  setStage(inv2, "discovering", "completed");

  const pending = plan.acquisitions.filter(
    (a) => a.paymentState === "payment_required"
  );

  if (pending.length > 0) {
    if (createServerSigner()) {
      // Auto-acquire downstream from Inquvia's own wallet (real x402).
      await acquireDownstream(investigationId, userId);
      await finalizeInvestigation(investigationId);
      return (await getInvestigation(investigationId)) as Investigation;
    }
    inv2.status = "awaiting_payment";
    setStage(inv2, "awaiting_payment", "active");
    inv2.updatedAt = new Date().toISOString();
    await saveInvestigation(inv2);
    return inv2;
  }

  // Nothing left to acquire: run the capability's own analysis to completion.
  await finalizeInvestigation(investigationId);
  return (await getInvestigation(investigationId)) as Investigation;
}

/**
 * Finalize a capability investigation: run the capability-specific analyzer
 * (or the caller-supplied one, or the legacy heuristic), write the assessment,
 * and complete the investigation. Only acquires/evidence that actually settled
 * count towards the result — nothing is fabricated.
 */
export async function finalizeInvestigation(
  id: string,
  analyze?: CapabilityAnalyzer
): Promise<Investigation> {
  const inv = await getInvestigation(id);
  if (!inv) throw new Error("Investigation not found");

  const acqs = inv.acquisitions ?? [];
  const unresolved = acqs.filter((a) => !RESOLVED_STATES.includes(a.paymentState));

  // Still waiting on user to authorize/acquire evidence.
  if (unresolved.length > 0) {
    return inv;
  }

  const evidence = (inv.evidence ?? []).filter((e) => e.status === "collected");
  const analyzer = analyze ?? getAnalyzer(inv.capability) ?? heuristicAnalysis;

  // Without a capability analyzer and without evidence, completing would mean
  // fabricating an assessment — refuse and report the honest state instead.
  if (evidence.length === 0 && !getAnalyzer(inv.capability)) {
    inv.status = "evidence_unavailable";
    inv.conclusion = "insufficient_evidence";
    inv.conclusionText =
      "Insufficient evidence was acquired to render an assessment. The investigation could not be completed because no paid external evidence was obtained.";
    inv.confidence = 0;
    inv.limitations = [
      "No external evidence was successfully acquired",
      inv.blockReason ?? "Evidence acquisition did not complete",
    ];
    inv.updatedAt = new Date().toISOString();
    await saveInvestigation(inv);
    return inv;
  }

  // ANALYZING
  inv.status = "analyzing";
  setStage(inv, "analyzing", "active");
  await saveInvestigation(inv);
  const analysis = await analyzer(inv, evidence);

  // CROSS-CHECKING
  inv.status = "cross_checking";
  setStage(inv, "analyzing", "completed");
  setStage(inv, "cross_checking", "active");
  inv.supportingEvidenceIds = evidence
    .filter((e) => e.signal === "supporting" || e.supportsClaim)
    .map((e) => e.id);
  inv.contradictoryEvidenceIds = evidence
    .filter((e) => e.signal === "contradictory" || e.contradictsClaim)
    .map((e) => e.id);
  emit(inv, "cross_check_completed", "Cross-check of acquired evidence completed");
  await saveInvestigation(inv);

  // FINAL ASSESSMENT
  inv.conclusion = analysis.conclusion;
  inv.conclusionText = analysis.conclusionText;
  inv.confidence = analysis.confidence;
  inv.risk = analysis.risk;
  inv.findings = analysis.findings;
  inv.limitations = analysis.limitations;
  inv.contradictions = analysis.contradictions;
  inv.sourcesUsed =
    analysis.sourcesUsed?.length
      ? analysis.sourcesUsed
      : evidence.map((e) => e.source);
  if (analysis.uncertainty) inv.uncertainty = analysis.uncertainty;
  inv.evidenceGraph = buildEvidenceGraph(inv, evidence);
  inv.economicSummary = {
    totalSpend: acqs
      .filter((a) => a.paymentState === "evidence_received" || a.paymentState === "settled")
      .reduce((s, a) => s + a.amountMicro / 1e6, 0),
    checksPurchased: acqs.filter((a) => a.evidence).length,
    providerCategories: [
      ...new Set(acqs.filter((a) => a.evidence).map((a) => a.capability)),
    ],
    settlementStatus: "Settled",
    algorandRef: acqs.find((a) => a.txId)?.txId,
  };
  inv.status = "completed";
  setStage(inv, "cross_checking", "completed");
  setStage(inv, "assessment", "completed");
  inv.currentStage = undefined;
  inv.updatedAt = new Date().toISOString();
  emit(inv, "assessment_generated", "Assessment generated for capability investigation");
  await saveInvestigation(inv);
  return inv;
}

/**
 * Legacy generic investigation path (a single generic endpoint). Kept only for
 * backward compatibility — new independent capabilities use
 * startCapabilityInvestigation + planCapability + discoverAndAcquire.
 */
export async function runInvestigation(
  request: InvestigationRequest
): Promise<Investigation> {
  const id = request.id ?? `case_${nanoid(10)}`;
  let inv = await getInvestigation(id);

  if (!inv) {
    inv = createInvestigationRecord({
      id,
      userId: request.userId,
      title: "Investigation in Progress",
      question: request.question,
      inputs: request.inputs,
      inputType: detectInputType(request.inputs),
      status: "created",
    });
    inv.stages = initStages();
    await saveInvestigation(inv);
    emit(inv, "investigation_created", "Investigation created");
  }

  if (inv.status === "created" || inv.status === "planning") {
    const requirements = planEvidenceRequirements(
      inv.question,
      inv.inputs.map((i) => i.type)
    );
    inv = await planCapability(inv, requirements);
  }

  if (inv.status === "discovering") {
    inv = await discoverAndAcquire(id, request.userId ?? "");
  }

  return (await getInvestigation(id)) as Investigation;
}

/** Legacy alias used by some callers. */
export async function getInvestigationProgress(
  id: string
): Promise<Investigation | null> {
  return getInvestigation(id);
}

export { INVESTIGATION_BLOCKED_STATES };
export type { AnalysisResult };